### ISSUE TITLE
similarity transform matrix in c++ is different from python

### ISSUE THREAD

**Issue by twmht** [deepinsight/insightface#481]
given the landmark

```
[[35.916126 43.42934 ]
 [86.2129   49.016266]
 [55.2765   71.023384]
 [32.937046 87.26657 ]
 [72.59505  91.47491 ]]
```

and the python code

```
tform = trans.SimilarityTransform()
M=tform.estimate(dst, src)
print (M)
```

the matrix is

```
[[ 0.80088555  0.12822719  1.92989341]
 [-0.12822719  0.80088555 24.34255532]
 [ 0.          0.          1.        ]]
```

but in c++ code (from https://github.com/deepinsight/insightface/blob/master/cpp-align/FacePreprocess.h)

```
  float v1[5][2] = {
      {30.2946f, 51.6963f},
      {65.5318f, 51.5014f},
      {48.0252f, 71.7366f},
      {33.5493f, 92.3655f},
      {62.7299f, 92.2041f}};

  cv::Mat src(5,2,CV_32FC1, v1);

  memcpy(src.data, v1, 2 * 5 * sizeof(float));

  float v2[5][2] = 
  {{35.916126, 43.42934 },
   {86.2129 , 49.016266},
   {55.2765 , 71.023384},
   {32.937046,87.26657 },
   {72.59505, 91.47491 }};

  cv::Mat dst(5,2,CV_32FC1, v2);

  memcpy(dst.data, v2, 2 * 5 * sizeof(float));
  cv::Mat m = FacePreprocess::similarTransform(src, dst);

  std::cout << m << std::endl;

```

what I got is 

```
[1.1905804, -0.19061996, 13.114243;
  0.19061995, 1.1905805, -26.316315;
  0, 0, 1]
```

and apparently the image alignment result is wrong in c++.
 
Any idea?

**Comment by szad670401** [comment]
To swap src and dst.
`  cv::Mat m = FacePreprocess::similarTransform(dst ,src);
`

**Comment by twmht** [comment]
@szad670401 

I have tried that but there is one wrong element in the matrix

[[ 0.80088555  0.12822719  -6.0701103]
 [-0.12822719  0.80088555 24.34255532]
 [ 0.          0.          1.        ]]

**Comment by szad670401** [comment]
I will fix this problem next week.

**Comment by twmht** [comment]
@szad670401 

Any update?

**Comment by twmht** [comment]
@szad670401 

do you have time to solve this issue? if not, I can open a pull request for that.

**Comment by nttstar** [comment]
@twmht You can open a PR if you have already fixed it, thank you :)

**Comment by szad670401** [comment]
I try to solve this problem. But I find the result is normal ,the result of python is same as cpp one.

```
from skimage import transform as trans
import numpy as np

dst = np.array([[35.916126,43.42934 ],
 [86.2129  ,49.016266],
 [55.2765  ,71.023384],
 [32.937046,87.26657 ],
 [72.59505 ,91.47491 ]],dtype=np.float32)

src = np.array([
    [30.2946, 51.6963],
    [65.5318, 51.5014],
    [48.0252, 71.7366],
    [33.5493, `92.3655],`
    [62.7299, 92.2041]], dtype=np.float32)

tform = trans.SimilarityTransform()
tform.estimate(dst, src)
M = tform.params
print(M)

```
The output result which same as the cpp one.

```
[[  0.80088554   0.12822716  -6.07010382]
 [ -0.12822716   0.80088554  24.34255465]
 [  0.           0.           1.        ]]



```

**Comment by twmht** [comment]
@szad670401 

my mistake. it's really the +8 problem.

**Comment by dexception** [comment]
@twmht 
Can you share the final method ?

**Comment by jamessmith90** [comment]
@twmht 
Please update

**Comment by xiakj** [comment]
mark.

**Comment by xiakj** [comment]
@twmht whether need src[:,1]-=8

**Comment by cannguyen275** [comment]
Hey guys, sorry for bothering you. But my issue might be related to this issue.

That is I sometimes received an up-side-down aligned face when I used the C++ version of the face align code. Especially, it's usually happened with high face's yaw value.

Debugging brings me to a conclusion that it's an error in calculating the rotation value in SimilarityTransform in C++, when its scikit-image version in Python works perfectly without any issues.

I've created an issue in #1559 but no one care, then I try to re-post this issue around to find if someone can help.

Thanks a lot for your attention.
